src/posix/platform/system.cpp

```diff
@@ -303,14 +326,31 @@ void platformDeinit(void)
     return;
 }
 
+void platformDeinitOffloadMode(void)
+{
+#if OPENTHREAD_POSIX_VIRTUAL_TIME
+    virtualTimeDeinit();
+#endif
+    platformSpinelManagerDeinit();
+    sCoprocessorType = OT_COPROCESSOR_UNKNOWN;
+}
+
 void otSysDeinit(void)
 {
-    OT_ASSERT(gInstance != nullptr);
+    if (sCoprocessorType == OT_COPROCESSOR_RCP)
+    {
+        OT_ASSERT(gInstance != nullptr);
+        platformTearDown();
+        otInstanceFinalize(gInstance);
+        gInstance = nullptr;
```

Comment:
Why not putting this in `platformDeinitCoreMode`?

Reply:
I think it's somewhat interesting here. In posix platform init/deinit, there are 3 parts:
1. platformInit <-> platformDeinit
2. otInstanceInitSingle <-> otInstanceFinalize
3. platformSetup <-> platformTeardown

However when it's NCP, there is no 2. For 3., it doesn't exist at this stage. So this makes the NCP and RCP case look asymmetric in `otSysDeinit`. But I think the structure is good and we should still keep it even after we add NCP case.

So back to the question itself, it's because `platformTearDown` and `otInstanceFinalize` are peer steps of `platformDeinit`. So they shouldn't be put into it.